### PR TITLE
Live ranklist and rating delta prediction

### PR DESCRIPTION
**Pull request by meooow25**
### Changes
- Some modifications to the Codeforces API objects.
- A new `Ranklist` class to process standing rows from the API.
- Renamed `FutureContests` to `Contests` and added a ranklist command.

### What it looks like
ranklist

### CODE REVIEW

File: tle/util/cache_system2.py [hunk 1]
```diff
@@ -458,8 +485,18 @@ def __init__(self, cache_master):
...
-                get_current_rating = self.cache_master.rating_changes_cache.get_current_rating_for_handle
-                ranklist_by_contest[contest.id] = Ranklist(contest, problems, standings, get_current_rating, now)
+                get_current_rating = self.cache_master.rating_changes_cache.get_current_rating_or_default
+                if cf_common.is_nonstandard_contest(contest):
+                    ranklist = Ranklist(contest, problems, standings, now, get_current_rating, is_rated=False)
+                else:
+                    rated_range = None
+                    if 'Educational' in contest.name:
+                        # For some reason educational contests return all contestants in ranklist even
+                        # when unofficial contests are not requested.
```
Comment (krofna) [inline tle/util/cache_system2.py]: everyone is considered official but its still unrated for >2100 :D

Comment (meooow25) [reply tle/util/cache_system2.py]: That's how ~~Mafia~~ Codeforces works

File: tle/cogs/contests.py [hunk 2]
```diff
@@ -276,6 +286,177 @@ def _make_pages(self):
...
+        body_style = '{:>} {:<}    {:>}  ' + '  '.join(['{:>}'] * len(problem_indices))
+        header = ['#', 'Handle', '='] + problem_indices
+        if deltas:
+            header_style += '  {:^}'
+            body_style += '  {:>}'
+            header += ['\N{INCREMENT}']
+
+        body = []
+        for handle, standing in handle_standings:
+            tokens = [standing.rank, handle + ':', int(standing.points)]
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/cogs/contests.py]: Yes, that's the only reason. The usernames get highlighted in a different color.

File: tle/cogs/contests.py [hunk 3]
```diff
@@ -276,6 +286,177 @@ def _make_pages(self):
             await ctx.author.remove_roles(role, reason='User unsubscribed from contest reminders')
             await ctx.send(embed=discord_common.embed_success('Successfully unsubscribed from contest reminders'))
 
+    def _get_handles_of_guild_members(self, guild):
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/cogs/contests.py]: That seems reasonable. Currently the bot stores user information without any regard for guilds. Although it seems unlikely to be run in multiple guilds at once, I hope we can improve the database schema for the sake of it.

File: tle/cogs/contests.py [hunk 4]
```diff
@@ -276,6 +286,177 @@ def _make_pages(self):
...
+        if deltas:
+            for tokens, delta in zip(body, deltas):
+                tokens.append('' if delta is None else f'{delta:+}')
+        return header_style, body_style, header, body
+
+    def _make_standings_pages(self, contest, problem_indices, handle_standings, deltas=None):
+        pages = []
+        count = len(handle_standings)
+        handle_standings_chunks = [handle_standings[i: i + _STANDINGS_PER_PAGE]
+                                   for i in range(0, len(handle_standings), _STANDINGS_PER_PAGE)]
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/cogs/contests.py]: Agreed. Done.

File: tle/cogs/contests.py [hunk 5]
```diff
@@ -276,6 +286,177 @@ def _make_pages(self):
...
+        num_chunks = len(handle_standings_chunks)
+        if deltas:
+            delta_chunks = [deltas[i: i + _STANDINGS_PER_PAGE]
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/cogs/contests.py]: Done.

File: tle/cogs/contests.py [hunk 6]
```diff
@@ -276,6 +286,177 @@ def _make_pages(self):
             await ctx.author.remove_roles(role, reason='User unsubscribed from contest reminders')
             await ctx.send(embed=discord_common.embed_success('Successfully unsubscribed from contest reminders'))
 
+    def _get_handles_of_guild_members(self, guild):
+        # TODO: Change database to store users on a guild basis
+        res = cf_common.user_db.getallhandleswithrating()
+        return [handle for user_id, handle, rating in res]
+
+    @staticmethod
+    def _get_cf_standings_table(problem_indices, handle_standings, deltas=None):
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/cogs/contests.py]: Implemented, but a little differently. Let me know what you think.

**Pull request merged**